File: my-reader-mobile/src/services/core/catalog.ts
import type {
  BookDetail,
  CalibreBook,
  PaginatedBooks,
} from "@my-reader/tools/types/book"
import type { Library } from "@my-reader/tools/types/library"
import {
  type BookFormat,
  type BookContent,
  type BookSummary,
  type BookEntry as CoreBookEntry,
  catalogCountBooks,
  catalogCountLibraryBooks,
  catalogDeleteLocalBook,
  catalogGetBookDetail,
  catalogGetBookFormat,
  catalogGetLibraryBookDetail,
  catalogGetLibraryBookFormat,
  catalogGetLibraryIdentity,
  catalogGetLibraryUuid,
  catalogGetMyreaderBookContent,
  catalogImportLocalBook,
  catalogStageRemoteBookImport,
  catalogListBookFormats,
  catalogListBookSummaries,
  catalogListBooks,
  catalogListBooksPage,
  catalogListBooksPageByLastRead,
  catalogListLibraryBookFormats,
  catalogListLibraryBookSummaries,
  catalogListLibraryBooks,
  catalogListLibraryBooksPage,
  catalogListLibraryBooksPageByLastRead,
  catalogListLibrarySeriesBooks,
  catalogListSeriesBooks,
  catalogUpdateLocalBookMetadata,
  catalogValidateLibrary,
  contentGetFileState,
} from "my-reader-core"
import { appConfigPath } from "./app-config"
import { announceLocalSidecarWork } from "./sync-events"
import { toNativeFilesystemPath } from "../fs/path"
import { cacheFileState } from "../query/invalidate-table"
import type { FileState } from "./content"

export type CalibreBookSummary = BookSummary

export type CalibreBookFormat = BookFormat

function nativePath(libraryRootUri: string): string {
  return toNativeFilesystemPath(libraryRootUri)
}

function bookFromCore(book: CoreBookEntry): CalibreBook {
  return {
    ...book,
    series: book.series ?? null,
    seriesIndex: book.seriesIndex ?? null,
    timestamp: book.timestamp ?? null,
    pubdate: book.pubdate ?? null,
    lastModified: book.lastModified ?? null,
    comment: book.comment ?? null,
    publisher: book.publisher ?? null,
    preferredFormat: book.preferredFormat ?? null,
    rating: book.rating ?? null,
    uuid: book.uuid ?? null,
  }
}

export function validateCalibreLibrary(libraryRootUri: string): boolean {
  return catalogValidateLibrary(nativePath(libraryRootUri))
}

export function countCalibreBooks(libraryRootUri: string): Promise<number> {
  return catalogCountBooks(nativePath(libraryRootUri))
}

export async function listCalibreBooks(
  libraryRootUri: string,
): Promise<CalibreBook[]> {
  return (await catalogListBooks(nativePath(libraryRootUri))).map(bookFromCore)
}

export async function listCalibreBooksPage(
  libraryRootUri: string,
  offset: number,
  limit: number,
  sortBy?: string,
  search?: string,
): Promise<PaginatedBooks> {
  const page = await catalogListBooksPage(
    nativePath(libraryRootUri),
    offset,
    limit,
    sortBy,
    search,
  )
  return { items: page.items.map(bookFromCore), total: page.total }
}

export async function listCalibreBooksPageByLastRead(
  libraryRootUri: string,
  sidecarRootUri: string,
  offset: number,
  limit: number,
  search?: string,
): Promise<PaginatedBooks> {
  const page = await catalogListBooksPageByLastRead(
    nativePath(libraryRootUri),
    nativePath(sidecarRootUri),
    offset,
    limit,
    search,
  )
  return { items: page.items.map(bookFromCore), total: page.total }
}

export async function getCalibreBookDetail(
  libraryRootUri: string,
  bookId: number,
): Promise<BookDetail> {
  const book = await catalogGetBookDetail(nativePath(libraryRootUri), bookId)
  return {
    ...bookFromCore(book),
    titleSort: book.titleSort,
    formatSizes: book.formatSizes,
    identifiers: book.identifiers,
  }
}

export async function listCalibreSeriesBooks(
  libraryRootUri: string,
  seriesName: string,
  excludeBookId?: number,
): Promise<CalibreBook[]> {
  return (
    await catalogListSeriesBooks(
      nativePath(libraryRootUri),
      seriesName,
      excludeBookId,
    )
  ).map(bookFromCore)
}

export function getCalibreLibraryUuid(libraryRootUri: string): Promise<string> {
  return catalogGetLibraryUuid(nativePath(libraryRootUri))
}

export async function listCalibreBookSummaries(
  libraryRootUri: string,
): Promise<CalibreBookSummary[]> {
  return catalogListBookSummaries(nativePath(libraryRootUri))
}

export async function listCalibreBookFormats(
  libraryRootUri: string,
  bookId: number,
): Promise<CalibreBookFormat[]> {
  return catalogListBookFormats(nativePath(libraryRootUri), bookId)
}

export async function getCalibreBookFormat(
  libraryRootUri: string,
  bookId: number,
  format: string,
): Promise<CalibreBookFormat | undefined> {
  return catalogGetBookFormat(nativePath(libraryRootUri), bookId, format)
}

type CatalogLibrary = Pick<Library, "id">

export function countLibraryBooks(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
): Promise<number> {
  return catalogCountLibraryBooks(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
  )
}

export async function listLibraryBooks(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
): Promise<CalibreBook[]> {
  return (
    await catalogListLibraryBooks(
      appConfigPath,
      library.id,
      nativePath(sidecarRootUri),
      nativePath(contentRootUri),
    )
  ).map(bookFromCore)
}

export async function listLibraryBooksPage(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  offset: number,
  limit: number,
  sortBy?: string,
  search?: string,
): Promise<PaginatedBooks> {
  const page = await catalogListLibraryBooksPage(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    offset,
    limit,
    sortBy,
    search,
  )
  return { items: page.items.map(bookFromCore), total: page.total }
}

export async function listLibraryBooksPageByLastRead(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  offset: number,
  limit: number,
  search?: string,
): Promise<PaginatedBooks> {
  const page = await catalogListLibraryBooksPageByLastRead(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    offset,
    limit,
    search,
  )
  return { items: page.items.map(bookFromCore), total: page.total }
}

export async function getLibraryBookDetail(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  bookId: number,
): Promise<BookDetail> {
  const book = await catalogGetLibraryBookDetail(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    bookId,
  )
  return {
    ...bookFromCore(book),
    titleSort: book.titleSort,
    formatSizes: book.formatSizes,
    identifiers: book.identifiers,
  }
}

export async function listLibrarySeriesBooks(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  seriesName: string,
  excludeBookId?: number,
): Promise<CalibreBook[]> {
  return (
    await catalogListLibrarySeriesBooks(
      appConfigPath,
      library.id,
      nativePath(sidecarRootUri),
      nativePath(contentRootUri),
      seriesName,
      excludeBookId,
    )
  ).map(bookFromCore)
}

export function getLibraryIdentity(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
): Promise<string> {
  return catalogGetLibraryIdentity(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
  )
}

export function listLibraryBookSummaries(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
): Promise<CalibreBookSummary[]> {
  return catalogListLibraryBookSummaries(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
  )
}

export function listLibraryBookFormats(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  bookId: number,
): Promise<CalibreBookFormat[]> {
  return catalogListLibraryBookFormats(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    bookId,
  )
}

export function getLibraryBookFormat(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  bookId: number,
  format: string,
): Promise<CalibreBookFormat | undefined> {
  return catalogGetLibraryBookFormat(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    bookId,
    format,
  )
}

export function getMyreaderBookContent(
  contentRootUri: string,
  sidecarRootUri: string,
  bookId: number,
  format: string,
): Promise<BookContent> {
  return catalogGetMyreaderBookContent(
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    bookId,
    format,
  )
}

export async function importLocalBook(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  input: {
    sourceFileUri: string
    sourceFileName?: string
    title?: string
    authors: string[]
    consumeSourceFile: boolean
  },
): Promise<CalibreBook> {
  const book = await catalogImportLocalBook(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    {
      sourceFilePath: nativePath(input.sourceFileUri),
      sourceFileName: input.sourceFileName,
      title: input.title,
      authors: input.authors,
      recordedAtMs: Date.now(),
      consumeSourceFile: input.consumeSourceFile,
    },
  )
  announceLocalSidecarWork(library.id)
  return bookFromCore(book)
}

export async function importRemoteBook(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  input: {
    sourceFileUri: string
    sourceFileName?: string
    title?: string
    authors: string[]
    consumeSourceFile: boolean
  },
): Promise<CalibreBook> {
  const book = await catalogStageRemoteBookImport(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    {
      sourceFilePath: nativePath(input.sourceFileUri),
      sourceFileName: input.sourceFileName,
      title: input.title,
      authors: input.authors,
      recordedAtMs: Date.now(),
      consumeSourceFile: input.consumeSourceFile,
    },
  )
  const format = book.formats[0]
  if (format) {
    const importedFormat = await catalogGetLibraryBookFormat(
      appConfigPath,
      library.id,
      nativePath(sidecarRootUri),
      nativePath(contentRootUri),
      book.id,
      format,
    )
    const state = importedFormat
      ? await contentGetFileState(
          nativePath(sidecarRootUri),
          importedFormat.relativePath,
        )
      : undefined
    if (state) {
      const cachedState: FileState = {
        ...state,
        localSha256: state.localSha256 ?? null,
        localSize: state.localSize ?? null,
        localMtime: state.localMtime ?? null,
      }
      cacheFileState(library.id, cachedState)
    }
  }
  return bookFromCore(book)
}

export async function updateLocalBookMetadata(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  input: { bookId: number; title: string; authors: string[] },
): Promise<CalibreBook> {
  const book = await catalogUpdateLocalBookMetadata(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    {
      bookId: input.bookId,
      title: input.title,
      authors: input.authors,
      recordedAtMs: Date.now(),
    },
  )
  announceLocalSidecarWork(library.id)
  return bookFromCore(book)
}

export async function deleteLocalBook(
  library: CatalogLibrary,
  contentRootUri: string,
  sidecarRootUri: string,
  bookId: number,
): Promise<void> {
  await catalogDeleteLocalBook(
    appConfigPath,
    library.id,
    nativePath(sidecarRootUri),
    nativePath(contentRootUri),
    bookId,
    Date.now(),
  )
  announceLocalSidecarWork(library.id)
}

File: my-reader-mobile/modules/my-reader-core/rust/src/catalog.rs
use std::path::Path;

use my_reader_core::api::catalog::CatalogService;

use crate::{
    types::{
        required_i64, required_usize, BookContent, BookDetail, BookEntry, BookFormat, BookSummary,
        ImportBookRequest, PaginatedBooks, UpdateBookMetadataRequest,
    },
    CoreFfiError,
};

#[uniffi::export]
pub fn catalog_validate_library(library_root_path: String) -> bool {
    CatalogService::validate_library(Path::new(&library_root_path))
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_count_library_books(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
) -> Result<f64, CoreFfiError> {
    Ok(CatalogService::count_registered_library_books(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
    )
    .await
    .map_err(CoreFfiError::from_core)? as f64)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_library_books(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
) -> Result<Vec<BookEntry>, CoreFfiError> {
    Ok(CatalogService::list_registered_library_books(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_library_books_page(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    offset: f64,
    limit: f64,
    sort_by: Option<String>,
    search: Option<String>,
) -> Result<PaginatedBooks, CoreFfiError> {
    Ok(CatalogService::list_registered_library_books_page(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        required_usize(offset, "offset")?,
        required_usize(limit, "limit")?,
        sort_by.as_deref(),
        search.as_deref(),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_library_books_page_by_last_read(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    offset: f64,
    limit: f64,
    search: Option<String>,
) -> Result<PaginatedBooks, CoreFfiError> {
    Ok(
        CatalogService::list_registered_library_books_page_by_last_read(
            Path::new(&config_path),
            &library_id,
            Path::new(&sidecar_root_path),
            Path::new(&content_root_path),
            required_usize(offset, "offset")?,
            required_usize(limit, "limit")?,
            search.as_deref(),
        )
        .await
        .map_err(CoreFfiError::from_core)?
        .into(),
    )
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_library_book_detail(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    book_id: f64,
) -> Result<BookDetail, CoreFfiError> {
    Ok(CatalogService::get_registered_library_book_detail(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        required_i64(book_id, "bookId")?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_library_series_books(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    series_name: String,
    exclude_book_id: Option<f64>,
) -> Result<Vec<BookEntry>, CoreFfiError> {
    Ok(CatalogService::list_registered_library_series_books(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        &series_name,
        exclude_book_id
            .map(|value| required_i64(value, "excludeBookId"))
            .transpose()?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_library_identity(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
) -> Result<String, CoreFfiError> {
    CatalogService::get_registered_library_identity(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
    )
    .await
    .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_library_book_summaries(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
) -> Result<Vec<BookSummary>, CoreFfiError> {
    Ok(CatalogService::list_registered_library_book_summaries(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_library_book_formats(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    book_id: f64,
) -> Result<Vec<BookFormat>, CoreFfiError> {
    Ok(CatalogService::list_registered_library_book_formats(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        required_i64(book_id, "bookId")?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_library_book_format(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    book_id: f64,
    format: String,
) -> Result<Option<BookFormat>, CoreFfiError> {
    Ok(CatalogService::get_registered_library_book_format(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        required_i64(book_id, "bookId")?,
        &format,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .map(Into::into))
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_myreader_book_content(
    sidecar_root_path: String,
    content_root_path: String,
    book_id: f64,
    format: String,
) -> Result<BookContent, CoreFfiError> {
    Ok(CatalogService::get_myreader_book_content(
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        required_i64(book_id, "bookId")?,
        &format,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_import_local_book(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    request: ImportBookRequest,
) -> Result<BookEntry, CoreFfiError> {
    Ok(CatalogService::import_local_book(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        request.try_into()?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_stage_remote_book_import(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    request: ImportBookRequest,
) -> Result<BookEntry, CoreFfiError> {
    Ok(CatalogService::stage_remote_book_import(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        request.try_into()?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_update_local_book_metadata(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    request: UpdateBookMetadataRequest,
) -> Result<BookEntry, CoreFfiError> {
    Ok(CatalogService::update_local_book_metadata(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        request.try_into()?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_delete_local_book(
    config_path: String,
    library_id: String,
    sidecar_root_path: String,
    content_root_path: String,
    book_id: f64,
    recorded_at_ms: f64,
) -> Result<(), CoreFfiError> {
    CatalogService::delete_local_book(
        Path::new(&config_path),
        &library_id,
        Path::new(&sidecar_root_path),
        Path::new(&content_root_path),
        required_i64(book_id, "bookId")?,
        required_i64(recorded_at_ms, "recordedAtMs")?,
    )
    .await
    .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_count_books(library_root_path: String) -> Result<f64, CoreFfiError> {
    Ok(CatalogService::count_books(Path::new(&library_root_path))
        .await
        .map_err(CoreFfiError::from_core)? as f64)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_books(library_root_path: String) -> Result<Vec<BookEntry>, CoreFfiError> {
    Ok(CatalogService::list_books(Path::new(&library_root_path))
        .await
        .map_err(CoreFfiError::from_core)?
        .into_iter()
        .map(Into::into)
        .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_books_page(
    library_root_path: String,
    offset: f64,
    limit: f64,
    sort_by: Option<String>,
    search: Option<String>,
) -> Result<PaginatedBooks, CoreFfiError> {
    Ok(CatalogService::list_books_page(
        Path::new(&library_root_path),
        required_usize(offset, "offset")?,
        required_usize(limit, "limit")?,
        sort_by.as_deref(),
        search.as_deref(),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_books_page_by_last_read(
    library_root_path: String,
    sidecar_root_path: String,
    offset: f64,
    limit: f64,
    search: Option<String>,
) -> Result<PaginatedBooks, CoreFfiError> {
    Ok(CatalogService::list_books_page_by_last_read(
        Path::new(&library_root_path),
        Path::new(&sidecar_root_path),
        required_usize(offset, "offset")?,
        required_usize(limit, "limit")?,
        search.as_deref(),
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_book_detail(
    library_root_path: String,
    book_id: f64,
) -> Result<BookDetail, CoreFfiError> {
    Ok(CatalogService::get_book_detail(
        Path::new(&library_root_path),
        required_i64(book_id, "bookId")?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_series_books(
    library_root_path: String,
    series_name: String,
    exclude_book_id: Option<f64>,
) -> Result<Vec<BookEntry>, CoreFfiError> {
    Ok(CatalogService::list_series_books(
        Path::new(&library_root_path),
        &series_name,
        exclude_book_id
            .map(|value| required_i64(value, "excludeBookId"))
            .transpose()?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_library_uuid(library_root_path: String) -> Result<String, CoreFfiError> {
    CatalogService::get_library_uuid(Path::new(&library_root_path))
        .await
        .map_err(CoreFfiError::from_core)
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_book_summaries(
    library_root_path: String,
) -> Result<Vec<BookSummary>, CoreFfiError> {
    Ok(
        CatalogService::list_book_summaries(Path::new(&library_root_path))
            .await
            .map_err(CoreFfiError::from_core)?
            .into_iter()
            .map(Into::into)
            .collect(),
    )
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_list_book_formats(
    library_root_path: String,
    book_id: f64,
) -> Result<Vec<BookFormat>, CoreFfiError> {
    Ok(CatalogService::list_book_formats(
        Path::new(&library_root_path),
        required_i64(book_id, "bookId")?,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .into_iter()
    .map(Into::into)
    .collect())
}

#[uniffi::export(async_runtime = "tokio")]
pub async fn catalog_get_book_format(
    library_root_path: String,
    book_id: f64,
    format: String,
) -> Result<Option<BookFormat>, CoreFfiError> {
    Ok(CatalogService::get_book_format(
        Path::new(&library_root_path),
        required_i64(book_id, "bookId")?,
        &format,
    )
    .await
    .map_err(CoreFfiError::from_core)?
    .map(Into::into))
}
